#!/usr/bin/env Rscript
args = commandArgs(trailingOnly=TRUE)

if (!require("BiocManager")) install.packages("BiocManager")
if (!require("VariantAnnotation")) BiocManager::install(c("VariantAnnotation"))
if (!require("StructuralVariantAnnotation")) BiocManager::install(c("StructuralVariantAnnotation"))
if (!require("stringr")) install.packages("stringr")

library(VariantAnnotation)
library(StructuralVariantAnnotation)
library(stringr)

#' Simple SV type classifier in https://github.com/PapenfussLab/gridss/blob/master/example/simple-event-annotation.R
simpleEventType <- function(gr) {
  pgr = partner(gr)
  return(ifelse(seqnames(gr) != seqnames(pgr), "CTX", # inter-chromosomosal
                ifelse(strand(gr) == strand(pgr), "INV",
                       ifelse(gr$insLen >= abs(gr$svLen) * 0.7, "INS", # TODO: improve classification of complex events
                              ifelse(xor(start(gr) < start(pgr), strand(gr) == "-"), "DEL",
                                     "DUP")))))
}

# test if there is at least one argument: if not, return an error
if (length(args)==0) {
  stop("At least one argument must be supplied (input file, output basename)", call.=FALSE)
} else if (length(args)==1) {
}

#part 1: filter, annotate VCF
vcf <- VariantAnnotation::readVcf(args[1])
#add column to VCF
info(header(vcf)) = unique(as(rbind(as.data.frame(info(header(vcf))), data.frame(
  row.names=c("SIMPLE_TYPE"),
  Number=c("1"),
  Type=c("String"),
  Description=c("Simple event type annotation based purely on breakend position and orientation."))), "DataFrame"))
#read VCF into StructuralVariantAnnotation package
gr <- breakpointRanges(vcf)
# gr <- gr[gr$FILTER == "PASS" & partner(gr)$FILTER == "PASS"] # Remove low confidence calls
#detect and classify SVs in gr
svtype <- simpleEventType(gr)
info(vcf)$SIMPLE_TYPE <- NA_character_
info(vcf[gr$sourceId])$SIMPLE_TYPE <- svtype
info(vcf[gr$sourceId])$SVLEN <- gr$svLen
#write VCF to folder
writeVcf(vcf, paste(args[2],".sv.annotated.vcf",sep = "")) # generated by example/gridss.sh
# 
# #part 2: summary table
# simplegr <- gr[simpleEventType(gr) %in% c("INS", "INV", "DEL", "DUP")]
# simplebed <- data.frame(
#   chrom=seqnames(simplegr),
#   # call the centre of the homology/inexact interval
#   start=as.integer((start(simplegr) + end(simplegr)) / 2),
#   end=as.integer((start(partner(simplegr)) + end(partner(simplegr))) / 2),
#   name=simpleEventType(simplegr),
#   length=simplegr$svLen,
#   score=simplegr$QUAL,
#   strand="."
# )
# # Just the lower of the two breakends so we don't output everything twice
# simplebed <- simplebed[simplebed$start < simplebed$end,]
# write.table(simplebed, paste(args[2],".sv.annotated.simple.txt",sep = ""), 
#             quote=FALSE, sep='\t', row.names=FALSE)